### ISSUE TITLE
"React is not defined" reported at running

### ISSUE THREAD

**Issue by frozenjackxp** [micro-frontends-demo/browse#9]
Hi, Cam Jackson
when i run your demo code, an error reported on my Chrome browser:
-------------------------------------
ReferenceError: React is not defined
react
/Users/hedongjie/Desktop/developer/micro-frontends/container-master/external "React":1
> 1 | module.exports = React;
View compiled
__webpack_require__
/Users/hedongjie/Desktop/developer/micro-frontends/container-master/webpack/bootstrap:723
  720 | };
  721 | 
  722 | // Execute the module function
> 723 | modules[moduleId].call(module.exports, module, module.exports, hotCreateRequire(moduleId));
      | ^  724 | 
  725 | // Flag the module as loaded
  726 | module.l = true;
View compiled
fn
http://localhost:3000/static/js/bundle.js:101:20
   98 | /******/ 				);
   99 | /******/ 				hotCurrentParents = [];
  100 | /******/ 			}
> 101 | /******/ 			return __webpack_require__(request);
      |                    ^  102 | /******/ 		};
  103 | /******/ 		var ObjectFactory = function ObjectFactory(name) {
  104 | /******/ 			return {
View source
Module../src/index.js
http://localhost:3000/static/js/bundle.js:23632:63
-------------------------------------
i'm a green hand of frontends, any suggests for this issue? 
OS: macOS catalina(Darwin)
Node version: 12.14.0
npm version: 6.13.4
thanks.

**Comment by locngoxuan** [comment]
@frozenjackxp: Remove ` "config-overrides-path": "node_modules/react-app-rewire-micro-frontends"` in `package.json`, then add `config-overrides.js` under `./` directory fix this issue.

```javascripts
#config-overrides.js
module.exports = function override(config, env) {
    return config;
}
```

**Comment by FergusMcGlynn** [comment]
@locngoxuan That does get this micro frontend working - however it causes it (the 'Browse' micro frontend) to use its own version of `react` and `react-dom` (as defined in `package.json` dependencies) rather than making use of the ones hosted by the 'Content' micro frontend. This is because it removes the `config.externals` section from the `react-app-rewire-micro-frontends` library (see https://github.com/camjackson/react-app-rewire-micro-frontends/blob/master/config-overrides.js)

The "Common content" section of the Micro Frontends article describes making use of a shared `react` and `react-dom` in each micro frontend rather than each one bundling their own. This should minimise each one's bundle size. It would be nice to get this micro frontend working while still making use of the common content.

**Comment by FergusMcGlynn** [comment]
In case this is a help to anyone else: I was running into the same problem as @frozenjackxp in the original post, and the solution turned out to be that I needed to be use the UMD (Universal Module Definition) version of `react` and `react-dom` and not the CJS (CommonJS) version. The latter don't work in the browser unless they're transpiled first, and because they fail to load then `react` will be undefined. 

I needed to download 
https://cdnjs.cloudflare.com/ajax/libs/react/16.13.0/umd/react.production.min.js
and 
https://cdnjs.cloudflare.com/ajax/libs/react-dom/16.13.0/umd/react-dom.production.min.js
instead of
https://cdnjs.cloudflare.com/ajax/libs/react/16.13.0/cjs/react.production.min.js
and
https://cdnjs.cloudflare.com/ajax/libs/react-dom/16.13.0/cjs/react-dom.production.min.js

**Comment by camjackson** [comment]
Hi everyone. As @FergusMcGlynn mentioned, the micro frontends are set up to externalise React from the compiled bundle, so that a single instance can be shared across all the different apps, reducing the number of bytes that users have to download. It's described [here](https://martinfowler.com/articles/micro-frontends.html#CommonContent).

I can't say for sure but I suspect the issue is that you aren't running the [content](https://github.com/micro-frontends-demo/content) app alongside the rest of the applications, as the README for this repo describes. The content app provides common content such as images, styles, data, and shared dependencies like React.

I hope that helps.

**Comment by charlie-maloney** [comment]
Hi @camjackson -- I ran into a similar issue and saw this thread. Does this mean that during development, we are using the minified production build of React? Is there a way to use the development build during development? 

Also -- loved your article on https://martinfowler.com/.

**Comment by camjackson** [comment]
Hi @charlie-maloney, it's been ages since I looked at this code but I'm pretty sure the shared content server should be serving up the dev build of react during development, and then bundling up the production build for production. Are you seeing something that would suggest otherwise?

And thanks 🙂

**Comment by gvkhirwadkar** [comment]
> 

@FergusMcGlynn  If you can tell where exactly you put those links for downloading React via CDN ? Is it in index.html under public folder ?

**Comment by FergusMcGlynn** [comment]
@gvkhirwadkar It's a long time since I looked at this - but I don't think I embedded the links to `react` and `react-dom` in the application anywhere, not even in `public/index.html` of this repo. If you look at the [content](https://github.com/micro-frontends-demo/content) repo, it has a folder called `content`, and in there, there are downloaded versions of `react` and `react-dom`. It is these downloaded copies of `react` and `react-dom` that I needed to replace with UMD versions rather than CJS versions. I did that simply by downloading the UMD files into the `content` folder (overwriting the ones that were already there) using the links in my comment above.

Bear in mind that those versions of `react` and `react-dom` will be out of date by now.
